File: sdks/python/src/caspian_sdk/state.py
"""Pluggable state and deduplication adapters for Caspian SDK."""

import logging
import uuid
from collections.abc import AsyncIterator
from contextlib import AbstractAsyncContextManager, asynccontextmanager
from threading import Lock
from typing import Any, Protocol, runtime_checkable

logger = logging.getLogger("caspian_sdk.state")

LUA_RELEASE_LOCK = """
if redis.call('get', KEYS[1]) == ARGV[1] then
    return redis.call('del', KEYS[1])
else
    return 0
end
"""


@runtime_checkable
class StateAdapter(Protocol):
    """Protocol defining the interface for Caspian state adapters."""

    async def seen(self, event_id: str) -> bool:
        """Atomic deduplication check.

        Returns True if `event_id` is new (and claims it). Returns False if it
        is a duplicate. Must not raise on duplicates.
        """
        ...

    def lock(self, conversation_id: str) -> AbstractAsyncContextManager[bool]:
        """Best-effort per-conversation lock yielding a bool (True if acquired, False if not).

        Must use an async context manager so locks cannot leak on exceptions.
        """
        ...


class InMemoryStateAdapter:
    """Default zero-config in-memory state adapter.

    Features bounded deduplication set (FIFO eviction) and thread-safe lazy per-conversation
    locks cleaned up when unused. Safe under multi-threaded execution.
    """

    def __init__(self, max_size: int = 10000) -> None:
        """Initialize in-memory state adapter.

        :param max_size: Maximum size of the deduplication set before oldest items are evicted.
        """
        if max_size <= 0:
            raise ValueError("max_size must be positive")
        self._max_size = max_size
        self._seen: dict[str, None] = {}
        self._seen_lock = Lock()
        self._locks: dict[str, Lock] = {}
        self._lock_ref_counts: dict[str, int] = {}
        self._locks_guard = Lock()

    async def seen(self, event_id: str) -> bool:
        """Atomic claim of an event_id across threads. Returns True if new, False if duplicate."""
        with self._seen_lock:
            if event_id in self._seen:
                return False
            if len(self._seen) >= self._max_size:
                # Evict oldest entry (dict retains insertion order)
                oldest = next(iter(self._seen))
                del self._seen[oldest]
            self._seen[event_id] = None
            return True

    @asynccontextmanager
    async def lock(self, conversation_id: str) -> AsyncIterator[bool]:
        """Per-conversation lock yielding True if acquired, False if already locked."""
        with self._locks_guard:
            if conversation_id not in self._locks:
                self._locks[conversation_id] = Lock()
                self._lock_ref_counts[conversation_id] = 0
            lock_obj = self._locks[conversation_id]
            self._lock_ref_counts[conversation_id] += 1

        acquired = lock_obj.acquire(blocking=False)

        try:
            yield acquired
        finally:
            if acquired:
                lock_obj.release()

            with self._locks_guard:
                self._lock_ref_counts[conversation_id] -= 1
                if self._lock_ref_counts[conversation_id] == 0:
                    self._locks.pop(conversation_id, None)
                    self._lock_ref_counts.pop(conversation_id, None)


class RedisStateAdapter:
    """Redis-backed state adapter for multi-instance / distributed deployments."""

    def __init__(
        self,
        redis_client: Any = None,
        url: str | None = None,
        seen_ttl: int = 86400,
        lock_ttl: int = 30,
    ) -> None:
        """Initialize Redis state adapter.

        :param redis_client: Pre-configured redis async client instance.
        :param url: Redis connection URL if `redis_client` is not provided.
        :param seen_ttl: Deduplication key expiration in seconds. Default 86400 (24h).
            Rationale: Channel providers (Slack, Discord, Telegram, WhatsApp) retry failed
            webhook deliveries up to 24 hours. 24h ensures robust dedup across retry windows
            without unbounded key growth in Redis.
        :param lock_ttl: Per-conversation lock TTL in seconds. Default 30s.
            Rationale: 30s is long enough for typical handler runtimes (including LLM calls),
            while short enough that if a worker process crashes, the conversation lock
            auto-expires quickly without deadlocking the conversation indefinitely.
        """
        if redis_client is not None:
            self._redis = redis_client
        else:
            try:
                import redis.asyncio as aioredis
            except ImportError as exc:
                raise ImportError(
                    "redis-py is required to use RedisStateAdapter. "
                    "Install it with 'pip install redis'."
                ) from exc
            if url is not None:
                self._redis = aioredis.from_url(url)
            else:
                self._redis = aioredis.Redis()

        self._seen_ttl = seen_ttl
        self._lock_ttl = lock_ttl

    async def seen(self, event_id: str) -> bool:
        """Atomic SET event:{event_id} 1 NX EX <ttl>."""
        key = f"event:{event_id}"
        res = await self._redis.set(key, "1", nx=True, ex=self._seen_ttl)
        return bool(res)

    @asynccontextmanager
    async def lock(self, conversation_id: str) -> AsyncIterator[bool]:
        """Atomic SET lock:{conversation_id} <token> NX EX <ttl> and Lua release."""
        key = f"lock:{conversation_id}"
        token = str(uuid.uuid4())
        res = await self._redis.set(key, token, nx=True, ex=self._lock_ttl)
        acquired = bool(res)

        try:
            yield acquired
        finally:
            if acquired:
                # Release via Lua script so we only delete if token matches (avoids releasing a
                # lock that expired and was re-acquired by another worker).
                # TTL acts as the safety net if network issues prevent lock release.
                try:
                    await self._redis.eval(LUA_RELEASE_LOCK, 1, key, token)
                except Exception:
                    logger.exception(
                        "Failed to release Redis lock for conversation %s; "
                        "TTL will clean it up",
                        conversation_id,
                    )

File: sdks/typescript/src/state.ts
import crypto from "node:crypto";

export interface LockHandle {
  acquired: boolean;
  release: () => Promise<void>;
}

export interface StateAdapter {
  /**
   * Atomic deduplication check.
   * Returns true if eventId is new (and claims it). Returns false if duplicate.
   */
  seen(eventId: string): Promise<boolean>;

  /**
   * Best-effort per-conversation lock.
   * Returns a LockHandle with acquired: true/false and a release() cleanup function.
   */
  lock(conversationId: string): Promise<LockHandle>;
}

export interface InMemoryStateAdapterOptions {
  /** Maximum size of the seen set before oldest entries are evicted. Default 10000. */
  maxSize?: number;
}

/**
 * Default zero-config in-memory state adapter with bounded deduplication set (FIFO eviction)
 * and per-conversation locks.
 */
export class InMemoryStateAdapter implements StateAdapter {
  private readonly maxSize: number;
  private readonly seenSet = new Set<string>();
  private readonly lockedConversations = new Set<string>();

  constructor(options: InMemoryStateAdapterOptions = {}) {
    const maxSize = options.maxSize ?? 10000;
    if (maxSize <= 0) {
      throw new Error("maxSize must be positive");
    }
    this.maxSize = maxSize;
  }

  async seen(eventId: string): Promise<boolean> {
    if (this.seenSet.has(eventId)) {
      return false;
    }
    if (this.seenSet.size >= this.maxSize) {
      const oldest = this.seenSet.keys().next().value;
      if (oldest !== undefined) {
        this.seenSet.delete(oldest);
      }
    }
    this.seenSet.add(eventId);
    return true;
  }

  async lock(conversationId: string): Promise<LockHandle> {
    if (this.lockedConversations.has(conversationId)) {
      return {
        acquired: false,
        release: async () => {},
      };
    }
    this.lockedConversations.add(conversationId);
    return {
      acquired: true,
      release: async () => {
        this.lockedConversations.delete(conversationId);
      },
    };
  }
}

const LUA_RELEASE_LOCK = `
if redis.call('get', KEYS[1]) == ARGV[1] then
    return redis.call('del', KEYS[1])
else
    return 0
end
`;

export interface RedisStateAdapterOptions {
  /** Deduplication key expiration in seconds. Default 86400 (24h). */
  seenTtl?: number;
  /** Per-conversation lock TTL in seconds. Default 30s. */
  lockTtl?: number;
}

/**
 * Redis-backed state adapter for multi-instance / distributed deployments.
 */
export class RedisStateAdapter implements StateAdapter {
  private readonly redis: any;
  private readonly seenTtl: number;
  private readonly lockTtl: number;

  constructor(clientOrOptions?: any, options: RedisStateAdapterOptions = {}) {
    /**
     * seenTtl: Default 86400s (24h). Channel providers (Slack, Discord, Telegram, WhatsApp)
     * retry failed webhook deliveries up to 24 hours. 24h ensures robust dedup across retry
     * windows without unbounded key growth in Redis.
     *
     * lockTtl: Default 30s. Long enough for typical agent handler runtimes (including LLM calls),
     * while short enough that if a worker process crashes, the conversation lock auto-expires
     * quickly without deadlocking the conversation indefinitely.
     */
    this.seenTtl = options.seenTtl ?? 86400;
    this.lockTtl = options.lockTtl ?? 30;

    if (clientOrOptions && typeof clientOrOptions.set === "function") {
      this.redis = clientOrOptions;
    } else {
      try {
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        const RedisModule = require("ioredis");
        const Redis = RedisModule.default || RedisModule;
        this.redis = new Redis(clientOrOptions);
      } catch (err) {
        throw new Error(
          "ioredis is required to use RedisStateAdapter. Install it with 'npm install ioredis'.",
        );
      }
    }
  }

  async seen(eventId: string): Promise<boolean> {
    const key = `event:${eventId}`;
    const res = await this.redis.set(key, "1", "NX", "EX", this.seenTtl);
    return res === "OK";
  }

  async lock(conversationId: string): Promise<LockHandle> {
    const key = `lock:${conversationId}`;
    const token =
      typeof crypto.randomUUID === "function"
        ? crypto.randomUUID()
        : crypto.randomBytes(16).toString("hex");
    const res = await this.redis.set(key, token, "NX", "EX", this.lockTtl);
    const acquired = res === "OK";

    if (!acquired) {
      return {
        acquired: false,
        release: async () => {},
      };
    }

    return {
      acquired: true,
      release: async () => {
        try {
          await this.redis.eval(LUA_RELEASE_LOCK, 1, key, token);
        } catch {
          /* TTL safety net if network release fails */
        }
      },
    };
  }
}
